Ensure device gets FF theming

Steve-Mcl opened the pull request:
<details open>
  <summary>Update 2023/07/05</summary>

## Description

1. Adds dependency on `@flowforge/flowforge-nr-theme` (the theme project)
2. Adds function `writeThemeFiles` to `launcher.js` to write the theme files to `node_modules/@flowforge/nr-theme` for the device instance to load up

This approach was taken to avoid having to copy the theme files into the repository (as was done in the previous PR).  
Additionally, as it is a dependency of the agent, theme files will always be available to the device regardless
of whether the device is prohibited from accessing the internet/NPM.

NOTE: `node_modules` is used due to NR <= 3.0.2 due to issue loading plugin packages 
* https://github.com/node-red/node-red/commit/ba22b07dcea0bf5bdbd93e25e58632733597925e
* https://github.com/node-red/node-red/pull/3867

</details>

<details closed>
  <summary>Update 2023/07/03</summary>

## Description

1. Adds dependency on `@flowforge/flowforge-nr-theme` (the theme project)
2. Adds function `writeThemeFiles` to `launcher.js` to write the theme files to `nodes/@flowforge/nr-theme` for the device instance to load up

This approach was taken to avoid having to copy the theme files into the repository (as was done in the previous PR).  
Additionally, as it is a dependency of the agent, theme files will always be available to the device regardless
of whether the device is prohibited from accessing the internet/NPM.

</details>

<details closed>
  <summary>Pre 2023/07/03 (Outdated)</summary>


## Description

1. Adds npm script `scripts/update-theme.mjs`
   1. This copies the necessary theme dist files from relative local `flowforge-nr-launcher` (where the theme project now resides)
   2. If working standalone alone or via CI/CD, downloads files from [github](https://github.com/flowforge/flowforge-nr-launcher)
2. Adds the current theme dist (result of running `npm run update-theme`)
4. Updated readme to list out the npm scripts available (including the newly added update-theme script)

### FYI
* Although this looks like a large commit, the majority of files are the result of executing the new script to copy in the latest theme dist.
   * The main changes are: 
      * Addition of `scripts/update-theme.mjs`: The script that brings latest theme files into the repository
      * Update of `template-settings.js`: adds `nodesDir: ['nodes']` to the device settings.js file - for the device instance to find the theme
      * Update of `launcher.js`: writes the theme files to `nodes/@flowforge/flowforge-nr-theme` for the device instance to load up
* At this point in time, only the light theme is used (necessary settings from the main project are not yet mapped through to device settings).  A separate issue will be opened in the FlowForge repo to cover this.
* This PR does not cover the CI/CD aspect (a separate issue will be raised to cover CI)


### Script output examples

#### Theme files found in relative path

```log
################################################################################
###                             Update theme files.                          ###
################################################################################
Checking local directory '../flowforge-nr-launcher' for theme files
Copying theme files from local directory
Writing README.md
Writing package.json
Theme files updated
################################################################################
```

#### Theme files downloaded from github

```log
################################################################################
###                             Update theme files.                          ###
################################################################################
Checking local directory '../flowforge-nr-launcher' for theme files
Local install not found.
Downloading files from 'https://github.com/flowforge/flowforge-nr-launcher'
Copying theme files from downloaded repository
Writing README.md
Writing package.json
Theme files updated
################################################################################
```

</details>


### Screenshots

#### Remote Editor - Before
![image](https://github.com/flowforge/flowforge-device-agent/assets/44235289/1fb0ecdf-e06f-4641-b981-89e6d713a665)


#### Remote Editor - After 
![image](https://github.com/flowforge/flowforge-device-agent/assets/44235289/671e1c4d-5ae0-47d6-bbf6-5f73815c175f)


## Related Issue(s)

#89

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass <!-- If not adding test coverage, please clarify why not? -->
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

hardillb commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 node_modules/
 package-lock.json
 var/
+temp/
```
Should this have been checked in?

Steve-Mcl replied:
yes, temp is generated by the script. It is cleaned up - but can get left behind if there is a local issue.

Steve-Mcl replied:
no longer valid since theme is now a dependency

hardillb commented on README.md:
```diff
@@ -197,3 +197,18 @@ The UI will only be available for the duration specified by the `--ui-runtime`.
+
+ - `npm start` - Start the agent
+ - `npm run dev` - Build the agent and watch for changes
+ - `npm run update-theme` - Update the theme files from the nr-launcher repository
```
This probably needs adding to the build build GH Action to ensure it always has the latest.

Also need to check the admin/build scripts and order in the release issue to ensure things are done in the right order.

Steve-Mcl replied:
it does and is raised as a task.

hardillb commented:
Does this need to duplicate files, can it not npm install the theme as a package?

Steve-Mcl commented:
> can it not npm install the theme as a package?

It could if the theme were a standalone package (but it is not).

(I am certain you are aware of that Ben, so I suspect I am reading your comment wrong)

hardillb commented:
> It could if the theme were a standalone package (but it is not).

I forgot it got merged into the launcher., I suspect it should be it's own repo and a sub-project for both this and nr-launcher so we don't have to copy files like this.

hardillb commented:
Can you also fix the conflict and I'll check it again.

Steve-Mcl commented:
> > It could if the theme were a standalone package (but it is not).
> 
> I forgot it got merged into the launcher., I suspect it should be it's own repo and a sub-project for both this and nr-launcher so we don't have to copy files like this.

Maybe. And [it was considered](https://github.com/flowforge/flowforge-device-agent/issues/89#issuecomment-1586192869) but...

Steve-Mcl commented:
> Can you also fix the conflict and I'll check it again.

Done @hardillb

hardillb commented:
I'm really not happy with copying files between repos, it's way too prone to errors.

The theme files should be converted to a git submodule and included in both repos (device-agent and nr-launcher).

Steve-Mcl commented:
@hardillb - I have removed the duplicate files and added a POSTINSTALL script.  This works in both the dev repo (`npm i` at top level) and when this package is installed as a standalone agent on a device.

Steve-Mcl commented:
> git installed

Ah. yes. since it downloads from Git, they will need git :( 

Boo!

If this is an issue, we could use the pure JS implementation?

Or we could "fail safe" (i.e. not install the theme)

Or we could grab the tarball from NPM.

Open to more suggestions (and beginning to wish we had just ripped the theme back out of launcher or launcher/agent were the same project :( )

hardillb commented:
Yeah, git is not included in the node:alpine-16 base container we use to build the device agent container.

Steve-Mcl commented:
back to the drawing board for now then.

Will reassess with fresh head next week.

Steve-Mcl commented:
Not too late to go back to 1st iteration or the PR ;)

Steve-Mcl commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
 node_modules/
 var/
+
```
```suggestion
```

revert

Steve-Mcl commented on README.md:
```diff
@@ -266,3 +266,18 @@ The UI will only be available for the duration specified by the `--ui-runtime`.
+
+ - `npm start` - Start the agent
+ - `npm run dev` - Build the agent and watch for changes
+ - `npm run update-theme` - Update the theme files from the nr-launcher repository
```
```suggestion
```

Remove script - no longer required

Steve-Mcl commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "scripts": {
         "start": "node index.js",
         "dev": "nodemon index.js",
+        "update-theme": "node ./scripts/update-theme.mjs",
```
```suggestion
```

Script no longer required (now, revering back to importing via npm package dependency)

Steve-Mcl commented:
@hardillb updated device agent.

* Removed all theme files
* Removed all scripts
* Added theme as dependency
* Added `writeThemeFiles` function - see OP for details

### IMPORTANT:

Will need to `npm i @flowforge/nr-theme` once the updated launcher (and sub module nr-theme) is published so that the package and package lock is updated to v0.2.0

Unsure if we can do this upfront?

Steve-Mcl commented on package.json:
```diff
@@ -29,6 +29,7 @@
     "author": "FlowForge Inc.",
     "license": "Apache-2.0",
     "dependencies": {
+        "@flowforge/nr-theme": "^0.1.3",
```
@hardillb - As discussed in Eng mtg (tagged here for quick reference)

We will need to `npm i @flowforge/nr-theme@latest` and push after a manual publish of theme versioned as 1.8.0

Steve-Mcl commented:
@hardillb - work is now complete & can be re-re-reviewed 🤞
